### PR TITLE
*: Add Row interface

### PR DESCRIPTION
**Pull request by coocood**
Add `Row` interface and make `[]types.Datum` implements it.
So we can change the implementation to Chunk Row in the future.

### CODE REVIEW

**Comment by coocood** [comment]
@zz-jason @XuHuaiyu PTAL

File: expression/column.go [hunk 1]
```diff
@@ -193,13 +193,13 @@ func (col *Column) GetType() *types.FieldType {
...
-	return row[col.Index], nil
+func (col *Column) Eval(row types.Row) (types.Datum, error) {
+	return row.(types.DatumnRow)[col.Index], nil
 }
 
 // EvalInt returns int representation of Column.
-func (col *Column) EvalInt(row []types.Datum, sc *variable.StatementContext) (int64, bool, error) {
-	val := &row[col.Index]
+func (col *Column) EvalInt(row types.Row, sc *variable.StatementContext) (int64, bool, error) {
+	val := &row.(types.DatumnRow)[col.Index]
```
Comment (shenli) [inline expression/column.go]: Will this kind of type assert be time consuming?

File: util/types/row.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+	GetMyDecimal(colIdx int) (*MyDecimal, bool)
+
+	// GetJSON returns the JSON value and isNull with the colIdx.
+	GetJSON(colIdx int) (json.JSON, bool)
+}
+
+var _ Row = DatumnRow{}
+
+// DatumnRow is a slice of Datum, implements Row interface.
+type DatumnRow []Datum
```
Comment (zz-jason) [inline util/types/row.go]: s/DatumnRow/DatumRow/

File: util/types/row.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
+// Copyright 2015 PingCAP, Inc.
```
Comment (lamxTyler) [inline util/types/row.go]: 2015 -> 2017

File: util/types/row.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+// Row is an interface to read columns values.
+type Row interface {
+
+	// GetInt64 returns the int64 value and isNull with the colIdx.
+	GetInt64(colIdx int) (val int64, isNull bool)
+
+	// GetUint64 returns the uint64 value and isNull with the colIdx.
+	GetUint64(colIdx int) (val uint64, isNull bool)
+	//
```
Comment (lamxTyler) [inline util/types/row.go]: Extra `//`?

File: util/types/row.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+	// GetFloat32 returns the float64 value and isNull with the colIdx.
```
Comment (lamxTyler) [inline util/types/row.go]: float64 -> float32

File: expression/column.go (hunk 1 above)
Comment (coocood) [reply expression/column.go]: It's just temporary change, this type assertion will be removed in the next PR.

**Comment by coocood** [comment]
@zz-jason @lamxTyler PTAL

**Comment by zz-jason** [comment]
LGTM

**Pull request merged**